src/pages/RegisterPage/components/RegisterForm/RegisterForm.tsx

```diff
@@ -1,42 +1,122 @@
-import { FC } from "react";
+import { FC, useState } from "react";
+import { useNavigate } from "react-router-dom";
 import { Formik, Form } from "formik";
 import cn from "classnames";
 
-import { UIbutton, TextField } from "@/src/components";
-import { IRegisterAccount, RegisterFormProps } from "./RegisterForm.type";
-
+import RegisterFormApi from "./RegisterFormApi";
+import {
+  type RegisterFormProps,
+  RegisterAccountValues,
+  SetFieldError,
+  ResetForm,
+  ErrorMessage,
+  Steps,
+  RegisterAccountKey,
+} from "./RegisterForm.type";
+import { validationSchema } from "./validationSchema";
+import { links } from "@/src/utils";
 import styles from "./RegisterForm.module.scss";
 
-const initialValues: IRegisterAccount = { email: "", signUpCode: "" };
+import { UIbutton, TextField } from "@/src/components";
+import FormNotification from "../FormNotification/FormNotification";
+
+const initialValues: RegisterAccountValues = {
+  email: "",
+  hash: "",
+};
 
 const RegisterForm: FC<RegisterFormProps> = ({ className, ...props }) => {
+  const [step, setStep] = useState(Steps.FIRST);
+
+  const navigate = useNavigate();
+
+  const isNextStep = step > Steps.FIRST;
+
+  const onHandleSubmit = async <
+    T extends RegisterAccountValues,
+    K extends SetFieldError,
+    N extends ResetForm,
+  >(
+    { email, hash }: T,
+    setFieldError: K,
+    resetForm: N
+  ) => {
+    if (step === Steps.SECOND) {
+      const { status, id } = await RegisterFormApi.fetchUserRegData({
+        hash,
+      });
+
+      return status === 404
+        ? setFieldError(RegisterAccountKey.HASH, ErrorMessage.HASH)
+        : navigate(`${links.ACCOUNT_CREATION}/${id}`);
+    }
+    const { error, status } = await RegisterFormApi.fetchUserRegData({
+      email,
+    });
+
+    if (
+      status === 422 &&
+      RegisterFormApi.formatErrorResponse(error) === "inactive"
+    ) {
+      resetForm({ values: { email, hash } });
+      return setStep(step + 1);
+    }
+
+    if (status === 422)
+      return setFieldError(RegisterAccountKey.EMAIL, ErrorMessage.EMAIL);
+
+    resetForm({ values: { email, hash } });
+    setStep(step + 1);
+  };
+
+  const renderNextStep = (value: string) =>
+    step > Steps.FIRST ? (
+      <>
+        <FormNotification />
+        <TextField
+          id={RegisterAccountKey.HASH}
+          name={RegisterAccountKey.HASH}
+          dataAutomation={`${RegisterAccountKey.HASH}Input`}
+          label="Sign up code"
+          value={value}
+        />
+      </>
+    ) : null;
+
   return (
     <Formik
       initialValues={initialValues}
-      onSubmit={(values, { setSubmitting }) => {
-        setTimeout(() => {
-          console.log(values);
-          setSubmitting(false);
-        }, 1000);
+      validationSchema={validationSchema(isNextStep)}
+      onSubmit={async (
```

Comment:
could you please create separate function onSubmit instead this mess, after you should not create complicated function with 3 generic, check please params in props onSubmit in formik and create function with 2 params(values, formikHelpers)

Reply:
done